### PR TITLE
fix(invoke): Write in UTF-8 string instead of bytes.

### PR DESCRIPTION
**Pull request by jfuss**
It appears that we were using sys.stdout.buffer to support python2 and python3 at the same time. Switching to just write to sys.stdout allows us to write a utf-8 encoding string. When using sys.stdout.buffer, we can only write bytes and I couldn't get the correct UTF8 encoded string to print correctly.

#### Which issue(s) does this change fix?

#5229

#### Why is this change necessary?
Should allow us to write utf8 correctly to the terminal.

#### How does it address the issue?
See above git commit message

#### What side effects does this change have?
We may need to change how we write in more places but my system python is all messed up so going to rely on our tests to validate in the PR. I did validate this locally with the example in the Github Issue.

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [ ] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [ ] Write/update unit tests
- [ ] Write/update integration tests
- [ ] Write/update functional tests if needed
- [ ] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

**Review by hawflau — APPROVED** [review]
LGTM

File: samcli/lib/utils/stream_writer.py [hunk 1]
```diff
@@ -22,7 +22,7 @@ def __init__(self, stream, auto_flush=False):
     def stream(self):
         return self._stream
 
-    def write(self, output, encode=False):
+    def write(self, output, encode=False, write_to_buffer=True):
```
Comment (mndeveci) [inline samcli/lib/utils/stream_writer.py:25]: Just wondering, are there other places where this is called with `write_to_buffer=True`?

**Review by mndeveci — APPROVED** [review]

**Comment by jfuss** [comment]
Added integration test for this. I ran both of develop and my branch. Results below:

develop:
```
$ SAM_CLI_DEV=1 pytest tests/integration -k test_invoke_returns_utf8
...

tests/integration/local/invoke/test_integrations_cli.py::TestSamPython37HelloWorldIntegration_0::test_invoke_returns_utf8 FAILED                                                                                                                                                                                                                                     [ 50%]
tests/integration/local/invoke/test_integrations_cli.py::TestSamPython37HelloWorldIntegration_1::test_invoke_returns_utf8 FAILED
...
FAILED tests/integration/local/invoke/test_integrations_cli.py::TestSamPython37HelloWorldIntegration_0::test_invoke_returns_utf8 - AssertionError: '{"key1": "안녕", "key2": "😍", "key3": "ḧ̸͈̟o̸̡̧͒ẉ̴̌ ̶[67 chars]̛͈"}' != '{"key1": "\\uc548\\ub155", "key2": "\\ud83d\\ude0d"[512 chars]48"}'
FAILED tests/integration/local/invoke/test_integrations_cli.py::TestSamPython37HelloWorldIntegration_1::test_invoke_returns_utf8 - AssertionError: '{"key1": "안녕", "key2": "😍", "key3": "ḧ̸͈̟o̸̡̧͒ẉ̴̌ ̶[67 chars]̛͈"}' != '{"key1": "\\uc548\\ub155", "key2": "\\ud83d\\ude0d"[512 chars]48"}'
```

jfuss:issue/5229:
```
$ SAM_CLI_DEV=1 pytest tests/integration -k test_invoke_returns_utf8
...

tests/integration/local/invoke/test_integrations_cli.py::TestSamPython37HelloWorldIntegration_0::test_invoke_returns_utf8 PASSED                                                                                                                                                                                                                                     [ 50%]
tests/integration/local/invoke/test_integrations_cli.py::TestSamPython37HelloWorldIntegration_1::test_invoke_returns_utf8 PASSED
```

File: samcli/lib/utils/stream_writer.py (hunk 1 above)
Comment (jfuss) [reply samcli/lib/utils/stream_writer.py:25]: I added `write_to_buffer` to keep all the previous code paths the same. I was seeing different behavior when writing to `sys.stdout.buffer` and `sys.stdout`. I wanted to write to `sys.stdout` for the output of the invoke but leave everything else writing to the buffer. So I added this option to the write call, defaulted to the previous behavior of writing to the buffer, and then passing is `False` for my specific case. 

Ideally, I would have wanted to change everything but that was a much bigger change and coming from all the docker upgrade issues, I went the least impactful route.

**Review by hawflau — APPROVED** [review]

**Pull request merged**